### ISSUE TITLE
ImportError: cannot import name 'CLOSED' from 'websockets.connection'

### ISSUE THREAD

**Issue by vcidst** [sanic-org/sanic#2733]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We're using Sanic `21.12.2` at rasa and notice this bug whenever rasa tries to spin a sanic server,

```
File "/Users/zi/Work/Rasa/venv-rasa-oss-3-10-2/lib/python3.10/site-packages/sanic/mixins/startup.py", line 57, in <module>
    from sanic.server.protocols.websocket_protocol import WebSocketProtocol
  File "/Users/zi/Work/Rasa/venv-rasa-oss-3-10-2/lib/python3.10/site-packages/sanic/server/protocols/websocket_protocol.py", line 3, in <module>
    from websockets.connection import CLOSED, CLOSING, OPEN
ImportError: cannot import name 'CLOSED' from 'websockets.connection' (/Users/zi/Work/Rasa/venv-rasa-oss-3-10-2/lib/python3.10/site-packages/websockets/connection.py)
```

It seems like https://github.com/sanic-org/sanic/pull/2609 addresses it but these changes are not available to the branch for version 21 `21.12LTS`. 

### Expected Behavior

I would have expected to not see any error when starting the sanic server. I get expected behaviour when using sanic version `22.12.0` and `23.3.0`

### How do you run Sanic?

Sanic CLI

### Operating System

MacOS

### Sanic Version

Sanic 21.12.2; Routing 0.7.2

**Comment by aaugustin** [comment]
I submitted changes to Sanic in November to prevent this issue:
https://github.com/sanic-org/sanic/commit/4c14910d5bbbbb6e12c1860572ecdc5c198f9b7e

If you're sticking to an old version of Sanic, you have to stick to an old version of websockets too, specifically `<11.0`.

Indeed, Sanic used to rely on implementation details of websockets, before I submitted that MR to use the public API.

**Comment by aaugustin** [comment]
If Sanic's stability policies require fixing it for 21.12 LTS, I would recommend adding the constraint `websockets<11.0` in the dependencies of that version.

Alternatively, it would be technically possible to address this in websockets by restoring aliases for the bits that Sanic uses. However, this sets a bad precedent for the distinction between the public API and the rest of the implementation.

**Comment by ahopkins** [comment]
No, we should have the constraint. That is absolutely what the intent was, and if it's not there, then I will add it. I'm away from my computer for a few days, so you will manually need to constrain your env.

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — this one looks closeable: the `from websockets.connection import CLOSED, CLOSING, OPEN` import that broke against newer `websockets` packages has been replaced with a version-tolerant try/except.

Evidence:
- [sanic/server/websockets/impl.py:14-19](https://github.com/sanic-org/sanic/blob/785d77f8fe208576a48d339bd81d866021154092/sanic/server/websockets/impl.py#L14-L19) now does:
  ```python
  try:  # websockets >= 11.0
      from websockets.protocol import Event, State  # type: ignore
      from websockets.server import ServerProtocol  # type: ignore
  except ImportError:  # websockets < 11.0
      from websockets.connection import Event, State  # type: ignore
      from websockets.server import ServerConnection as ServerProtocol
  ```
  and then derives `OPEN = State.OPEN`, `CLOSING = State.CLOSING`, `CLOSED = State.CLOSED` (L30-32) regardless of which branch the import took. This covers both the modern `websockets.protocol` location and the old `websockets.connection` location the original trace pointed at.
- The original `sanic/server/protocols/websocket_protocol.py` import the traceback flagged no longer touches `CLOSED`/`CLOSING`/`OPEN` — that file [imports only `WebSocketCommonProtocol` and async helpers](https://github.com/sanic-org/sanic/blob/785d77f8fe208576a48d339bd81d866021154092/sanic/server/protocols/websocket_protocol.py) now.
- Sanic 25.12.0 requires Python 3.10+ and pins `websockets` so this incompatibility window with the original 21.12.2 + `websockets` mix is no longer reachable.

If you're still hitting an `ImportError` from `websockets.connection` against a recent sanic + websockets combo, post the exact pip-list and I'll dig further. Otherwise would you mind closing this out?

---
*Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the file/PR references above were verified manually against the v25.12 release commit.*
